tools/webtransport/h3/webtransport_h3_server.py

```diff
@@ -0,0 +1,307 @@
+import asyncio
+import logging
+import os
+import threading
+import traceback
+from urllib.parse import urlparse
+from typing import Any, Dict, List, Optional, Tuple
+
+from aioquic.asyncio import QuicConnectionProtocol, serve
+from aioquic.h3.connection import H3_ALPN, H3Connection
+from aioquic.h3.events import H3Event, HeadersReceived, WebTransportStreamDataReceived, DatagramReceived
+from aioquic.quic.configuration import QuicConfiguration
+from aioquic.quic.connection import stream_is_unidirectional
+from aioquic.quic.events import QuicEvent, ProtocolNegotiated
+from aioquic.tls import SessionTicket
+from aioquic.quic.packet import QuicErrorCode
+"""
+A WebTransport over HTTP/3 server for testing.
+
+The server interprets the underlying protocols (WebTransport, HTTP/3 and QUIC)
+and passes events to a particular webtransport handler. From the standpoint of
+test authors, a webtransport handler is a Python script which contains some
+callback functions. See handler.py for available callbacks.
+"""
+
+SERVER_NAME = 'webtransport-h3-server'
+
+_logger = None
+_doc_root = ""
+
+
+class WebTransportH3Protocol(QuicConnectionProtocol):
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self._handler: Optional[WebTransportEventHandler] = None
+        self._http: Optional[H3Connection] = None
+
+    def quic_event_received(self, event: QuicEvent) -> None:
+        if isinstance(event, ProtocolNegotiated):
+            self._http = H3Connection(self._quic, enable_webtransport=True)
+
+        if self._http is not None:
+            for http_event in self._http.handle_event(event):
+                self._h3_event_received(http_event)
+
+    def _h3_event_received(self, event: H3Event) -> None:
+        if isinstance(event, HeadersReceived):
+            # Convert from List[Tuple[bytes, bytes]] to Dict[bytes, bytes]
```

Comment:
Can you comment that only the last header will be kept, if that's the intention, and otherwise assert that this doesn't happen?

Reply:
Yes, that's the intention. Added comment.